ext/pg_tuple.c

```diff
@@ -0,0 +1,435 @@
+#include "pg.h"
+
+static VALUE rb_cPG_Tuple;
+
+typedef struct {
+	/* PG::Result object this tuple was retrieved from.
+	 * Qnil when all fields are materialized.
+	 */
+	VALUE result;
+
+	/* Store the typemap of the result.
+	 * It's not enough to reference the PG::TypeMap object through the result,
+	 * since it could be exchanged after the tuple has been created.
+	 */
+	VALUE typemap;
+
+	/* Hash with maps field names to index into values[]
+	 * Shared between all instances retrieved from one PG::Result.
+	 */
+	VALUE field_map;
+
+	/* Row number within the result set. */
+	int row_num;
+
+	/* Materialized values. */
+	VALUE values[0];
+} t_pg_tuple;
+
+static void
+pg_tuple_gc_mark( t_pg_tuple *this )
+{
+	int i;
+
+	if( !this ) return;
+	rb_gc_mark( this->result );
+	rb_gc_mark( this->typemap );
+	rb_gc_mark( this->field_map );
+
+	for( i = 0; i < (int)RHASH_SIZE(this->field_map); i++ ){
+		rb_gc_mark( this->values[i] );
+	}
+}
+
+static void
+pg_tuple_gc_free( t_pg_tuple *this )
+{
+	if( !this ) return;
+	xfree(this);
+}
+
+static size_t
+pg_tuple_memsize( t_pg_tuple *this )
+{
+	if( this==NULL ) return 0;
+	return sizeof(*this) +  sizeof(*this->values) * RHASH_SIZE(this->field_map);
+}
+
+static const rb_data_type_t pg_tuple_type = {
+	"pg",
+	{
+		(void (*)(void*))pg_tuple_gc_mark,
+		(void (*)(void*))pg_tuple_gc_free,
+		(size_t (*)(const void *))pg_tuple_memsize,
+	},
+	0, 0,
+#ifdef RUBY_TYPED_FREE_IMMEDIATELY
+	RUBY_TYPED_FREE_IMMEDIATELY,
+#endif
+};
+
+/*
+ * Document-method: allocate
+ *
+ * call-seq:
+ *   PG::VeryTuple.allocate -> obj
+ */
+static VALUE
+pg_tuple_s_allocate( VALUE klass )
+{
+	return TypedData_Wrap_Struct( klass, &pg_tuple_type, NULL );
```

Comment:
I just learned about this way of writing extensions. Good to have a really clear example of how to do it.

Reply:
Thanks for reviewing Michael! The `Data_Wrap_Struct` way is actually [deprecated in the docs](https://github.com/ruby/ruby/blob/trunk/doc/extension.rdoc#encapsulate-c-data-into-a-ruby-object).

Reply:
Yeah, I've been meaning to convert all of my C extensions to use the new way, but it helps to see a working example. I just recently found a page called [The Definitive Guide to Ruby's C API](https://silverhammermba.github.io/emberb/c/), which had a bunch of stuff I didn't know or had forgotten as well.